src/btreemap.rs

```diff
@@ -81,10 +81,149 @@ const DEFAULT_PAGE_SIZE: u32 = 1024;
 // A marker to indicate that the `PageSize` stored in the header is a `PageSize::Value`.
 const PAGE_SIZE_VALUE_MARKER: u32 = u32::MAX;
 
-/// A "stable" map based on a B-tree.
+/// A B-Tree map implementation that stores its data into a designated memory.
 ///
-/// The implementation is based on the algorithm outlined in "Introduction to Algorithms"
-/// by Cormen et al.
+/// # Memory Implementations
+///
+/// `BTreeMap` works with any memory implementation that satisfies the [`Memory`] trait:
+///
+/// - [`Ic0StableMemory`](crate::Ic0StableMemory): Stores data in the Internet Computer's stable memory
+/// - [`VectorMemory`](crate::VectorMemory): In-memory implementation backed by a Rust `Vec<u8>`
+/// - [`FileMemory`](crate::FileMemory): Persists data to disk using a file
+/// - [`DefaultMemoryImpl`](crate::DefaultMemoryImpl): Default implementation that automatically selects the 
+///   appropriate memory backend based on the environment:
+///   - Uses `Ic0StableMemory` when running in an Internet Computer canister (wasm32 target)
+///   - Falls back to `VectorMemory` in other environments (like tests or non-IC contexts)
+///
+/// For almost all use cases, [`DefaultMemoryImpl`](crate::DefaultMemoryImpl) is recommended as it provides
+/// the right implementation based on the runtime context.
+///
+/// # Examples
+///
+/// ## Basic Usage with a Single BTreeMap
+///
+/// ```rust
+/// use ic_stable_structures::{BTreeMap, DefaultMemoryImpl};
+/// let mut map: BTreeMap<u64, String, _> = BTreeMap::init(DefaultMemoryImpl::default());
+///
+/// map.insert(1, "hello".to_string());
+/// # assert_eq!(map.get(&1), Some("hello".to_string()));
+/// ```
+///
+/// ## Multiple BTreeMaps and Memory Management
+///
+/// **Important**: Each stable structure requires its own designated memory region. Attempting to
+/// initialize multiple structures with the same memory will lead to data corruption.
+///
+/// ### What NOT to do:
+///
+/// ```rust,no_run
+/// use ic_stable_structures::{BTreeMap, DefaultMemoryImpl};
+/// 
+/// // ERROR: Using the same memory for multiple BTreeMaps will corrupt data
```

Comment:
Seems like this points out that the `DefaultMemoryImpl::default()` should internally do `memory_manager.get(MemoryId::new(next))`. This is not related to this PR, but just to confirm WDYT @ielashi ?

Reply:
Thanks for the suggestion @berestovskyy. I agree we should definitely think of ways of not requiring developers to set memory IDs manually. It's unfortunately not so straightforward. Where would you store this `next`? And how would you guarantee that these memory IDs are always the same? What if I add/remove structures between upgrades? We'd need to devise a mechanism that works across all these cases.

Reply:
I agree, it's not an easy change and should not be done in this PR. My high-level point is that to make the stable structures more user-friendly, we should avoid any default behaviors which might corrupt the data.

In this particular case, we first recommend the `DefaultMemoryImpl` for "almost all use cases", but a few lines later, we reveal that in fact for more than one stable structure we should use the `MemoryManager`.

Reply:
I see. The `MemoryManager` isn't a substitute for `DefaultMemoryImpl`. We should indeed use the `DefaultMemoryImpl` basically all the time, but it should be used _along with_ the MemoryManager. Maybe I can make this more clear in the docs.